source/common/config/delta_subscription_state.h

```diff
@@ -98,6 +100,8 @@ class DeltaSubscriptionState : public Logger::Loggable<Logger::Id::config> {
   // Feel free to change to unordered if you can figure out how to make it work.
   std::set<std::string> names_added_;
   std::set<std::string> names_removed_;
+  // Tracks resource aliases in on-demand requests since the previous DeltaDiscoveryRequest we sent.
+  std::set<std::string> requested_aliases_;
```

Comment:
I'm a bit confused; why are we resolving aliases at all? Don't we ask for a resource by name (a singular name), and the management server returns it, together with an alias set. We might ask for a resource by name later, and it might map into this alias set, but at no point does Envoy itself request aliases.. or maybe I'm not fully groking.

Reply:
I guess these are not really aliases (but rather route_config_name$host/authority-header), but they aren't resource names either (as they aren't opaque -- have a structure and a meaning). I thought they were closer to aliases. Ultimately the contents of the ```requested_aliases_``` field is copied into ```resource_names_subscribe``` of a DeltaDiscoveryRequest. I guess the name is very confusing, perhaps something like ```on_demand_queries``` would be better?

Reply:
Why can't they just be treated as normal resource subscription requests at this point?

Reply:
They can be (they are, in fact) the only difference is that I collect on-demand requests in a dedicated member var. If think this is excessive/using ```names_added_``` is not going to be confusing, I'll switch to using ```names_added_``` instead. @fredlas -- I don't think there are any downsides to that implementation-wise, are there?

Reply:
...actually, it goes beyond just the watches. DeltaSubscriptionState has to track a resource names->versions map. If VHDS ever uses its GrpcSubscriptionImpl to ask for both X and alias_of_X, you'll be tracking both of them, and only updating one or the other based on what name a resource arrives with. You could modify DeltaSubscriptionState to also go through the list of aliases when updating that map, but there's still a more fundamental question of what the whole xDS protocol wants: should the `initial_resource_versions` proto field be canonical names only?

Basically, if it's safe to say that an Envoy referring to a resource by an alias won't later start using the canonical name (or even a second alias), then everything can be simple and fine. If not, it will take some consideration and decisions.

Reply:
> One question - maybe already answered - is how you want the aliases to interact with the "watch" concept. Could you ever express interest in X, and then later express interest in alias_of_X?

I don't think there's a need or reason to track aliases, fwiw I'm tracking them now and wasn't planning on that. However, what I'll be adding to names_added_ isn't an alias or a resource name, but rather something like ```route_config_name$domain.name.org``` - a combination of a route config name and contents of a host header (still not tracking that!). 

I think I know the answer already, but to confirm -- DeltaSubscriptionState.names_added_ aren't directly involved in creation of watches, but are only used to create DeltaDiscoveryRequests. If so, it sounds like I should be ok.

Reply:
Hmm, in that case I think you'll end up with possibly meaningless (depending on how smart the server is) entries in `initial_resource_versions`. Which, maybe that's fine.

Regarding `names_added_` and watches, yes, you are correct, no influence. In fact, the opposite is true: the creation/destruction of watches is what (can) lead to changes in `names_added,removed_`.

And, because the influence flows in that direction, the cleanest thing, if possible, would be to just use `Subscription::updateResourceInterest` to express interest in the aliases. That's where using two names for the same logical resource would be an issue (solved by prompt cleanup when you're no longer interested in an alias).

Basically, it *seems* like it should be possible to do everything you need without modifying GrpcSubscription, GrpcMux, SubscriptionState. All of the interesting alias stuff happens in the response path, and moreover, mostly above the level of the xDS communication: the server notices a request for an alias, includes the alias in the resource it returns, and then VHDS (a user of the xDS client code) is what makes use of that alias field.

Reply:
>  don't think there's a need or reason to track aliases, fwiw I'm tracking them now and wasn't planning on that. 

Ugh, I meant to say I'm *not* tracking them now.

Reply:
> Hmm, in that case I think you'll end up with possibly meaningless (depending on how smart the server is) entries in initial_resource_versions. Which, maybe that's fine.

Could you expand on that a bit (or point to a place in the code if that's easier) pls?

> Basically, it seems like it should be possible to do everything you need without modifying GrpcSubscription, GrpcMux, SubscriptionState.

Nods, I think so too -- if on-demand requests aren't a special case, everything that was created to treat them as such can go.

Reply:
Sure, where it's happening is in `DeltaSubscriptionState::handleGoodResponse()`, when it calls `setResourceVersion()`. That updates the `resource_versions_` map, which if you take a look in `getNextRequestAckless()` gets copied into the "initial_resource_versions" proto field of the first request of every stream (i.e. reconnections).

...oh nice, actually, writing that out prompted me to think it through, and I think it's fine. Is it the case that the responses sent by the server will always set the resource's "name" field to be the canonical name, and the "btw it was resolved from this alias(es)" always just go in the "aliases" field? If that's the case, there is no problem.

Reply:
>  Is it the case that the responses sent by the server will always set the resource's "name" field to be the canonical name, and the "btw it was resolved from this alias(es)" always just go in the "aliases" field?

Yes, for responses to on-demand requests.

Reply:
Killed ```requested_aliases_``` field.

Reply:
Excellent! Much simpler. By the way, I'll be on vacation all of next week, in case you would be looking to discuss further.

Reply:
Enjoy your vacation! I'm sure this PR is still going to be around when you are back.